File: binding/include/printsupport/libqprintpreviewwidget.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qprintpreviewwidget.html
pub const qprintpreviewwidget = struct {

    /// New constructs a new QPrintPreviewWidget object.
    pub fn New(parent: ?*anyopaque) QtC.QPrintPreviewWidget {
        return qtc.QPrintPreviewWidget_new(@ptrCast(parent));
    }


    /// New2 constructs a new QPrintPreviewWidget object.
    pub fn New2(printer: ?*anyopaque) QtC.QPrintPreviewWidget {
        return qtc.QPrintPreviewWidget_new2(@ptrCast(printer));
    }


    /// New3 constructs a new QPrintPreviewWidget object.
    pub fn New3() QtC.QPrintPreviewWidget {
        return qtc.QPrintPreviewWidget_new3();
    }


    /// New4 constructs a new QPrintPreviewWidget object.
    pub fn New4(printer: ?*anyopaque, parent: ?*anyopaque) QtC.QPrintPreviewWidget {
        return qtc.QPrintPreviewWidget_new4(@ptrCast(printer), @ptrCast(parent));
    }


    /// New5 constructs a new QPrintPreviewWidget object.
    pub fn New5(printer: ?*anyopaque, parent: ?*anyopaque, flags: i64) QtC.QPrintPreviewWidget {
        return qtc.QPrintPreviewWidget_new5(@ptrCast(printer), @ptrCast(parent), @intCast(flags));
    }


    /// New6 constructs a new QPrintPreviewWidget object.
    pub fn New6(parent: ?*anyopaque, flags: i64) QtC.QPrintPreviewWidget {
        return qtc.QPrintPreviewWidget_new6(@ptrCast(parent), @intCast(flags));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrintPreviewWidget_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprintpreviewwidget.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ZoomFactor(self: ?*anyopaque, ) f64 {
        return qtc.QPrintPreviewWidget_ZoomFactor(@ptrCast(self));
    }

    pub fn Orientation(self: ?*anyopaque, ) i32 {
        return qtc.QPrintPreviewWidget_Orientation(@ptrCast(self));
    }

    pub fn ViewMode(self: ?*anyopaque, ) i32 {
        return qtc.QPrintPreviewWidget_ViewMode(@ptrCast(self));
    }

    pub fn ZoomMode(self: ?*anyopaque, ) i32 {
        return qtc.QPrintPreviewWidget_ZoomMode(@ptrCast(self));
    }

    pub fn CurrentPage(self: ?*anyopaque, ) i32 {
        return qtc.QPrintPreviewWidget_CurrentPage(@ptrCast(self));
    }

    pub fn PageCount(self: ?*anyopaque, ) i32 {
        return qtc.QPrintPreviewWidget_PageCount(@ptrCast(self));
    }

    pub fn SetVisible(self: ?*anyopaque, visible: bool) void {
        qtc.QPrintPreviewWidget_SetVisible(@ptrCast(self), visible);
    }

    /// Allows for overriding the related default method
    pub fn OnSetVisible(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QPrintPreviewWidget_OnSetVisible(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetVisible(self: ?*anyopaque, visible: bool) void {
        qtc.QPrintPreviewWidget_QBaseSetVisible(@ptrCast(self), visible);
    }

    pub fn Print(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_Print(@ptrCast(self));
    }

    pub fn ZoomIn(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_ZoomIn(@ptrCast(self));
    }

    pub fn ZoomOut(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_ZoomOut(@ptrCast(self));
    }

    pub fn SetZoomFactor(self: ?*anyopaque, zoomFactor: f64) void {
        qtc.QPrintPreviewWidget_SetZoomFactor(@ptrCast(self), zoomFactor);
    }

    pub fn SetOrientation(self: ?*anyopaque, orientation: i32) void {
        qtc.QPrintPreviewWidget_SetOrientation(@ptrCast(self), @intCast(orientation));
    }

    pub fn SetViewMode(self: ?*anyopaque, viewMode: i32) void {
        qtc.QPrintPreviewWidget_SetViewMode(@ptrCast(self), @intCast(viewMode));
    }

    pub fn SetZoomMode(self: ?*anyopaque, zoomMode: i32) void {
        qtc.QPrintPreviewWidget_SetZoomMode(@ptrCast(self), @intCast(zoomMode));
    }

    pub fn SetCurrentPage(self: ?*anyopaque, pageNumber: i32) void {
        qtc.QPrintPreviewWidget_SetCurrentPage(@ptrCast(self), pageNumber);
    }

    pub fn FitToWidth(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_FitToWidth(@ptrCast(self));
    }

    pub fn FitInView(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_FitInView(@ptrCast(self));
    }

    pub fn SetLandscapeOrientation(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_SetLandscapeOrientation(@ptrCast(self));
    }

    pub fn SetPortraitOrientation(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_SetPortraitOrientation(@ptrCast(self));
    }

    pub fn SetSinglePageViewMode(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_SetSinglePageViewMode(@ptrCast(self));
    }

    pub fn SetFacingPagesViewMode(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_SetFacingPagesViewMode(@ptrCast(self));
    }

    pub fn SetAllPagesViewMode(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_SetAllPagesViewMode(@ptrCast(self));
    }

    pub fn UpdatePreview(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_UpdatePreview(@ptrCast(self));
    }

    pub fn PaintRequested(self: ?*anyopaque, printer: ?*anyopaque) void {
        qtc.QPrintPreviewWidget_PaintRequested(@ptrCast(self), @ptrCast(printer));
    }

    pub fn OnPaintRequested(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QPrintPreviewWidget_Connect_PaintRequested(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn PreviewChanged(self: ?*anyopaque, ) void {
        qtc.QPrintPreviewWidget_PreviewChanged(@ptrCast(self));
    }

    pub fn OnPreviewChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QPrintPreviewWidget_Connect_PreviewChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrintPreviewWidget_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprintpreviewwidget.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrintPreviewWidget_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprintpreviewwidget.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ZoomIn1(self: ?*anyopaque, zoom: f64) void {
        qtc.QPrintPreviewWidget_ZoomIn1(@ptrCast(self), zoom);
    }

    pub fn ZoomOut1(self: ?*anyopaque, zoom: f64) void {
        qtc.QPrintPreviewWidget_ZoomOut1(@ptrCast(self), zoom);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPrintPreviewWidget_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/printpreviewwidget.html#types
pub const enums = struct {
    pub const ViewMode = enum {
        pub const SinglePageView: i32 = 0;
        pub const FacingPagesView: i32 = 1;
        pub const AllPagesView: i32 = 2;
    };

    pub const ZoomMode = enum {
        pub const CustomZoom: i32 = 0;
        pub const FitToWidth: i32 = 1;
        pub const FitInView: i32 = 2;
    };

};

File: binding/src/printsupport/libqprintpreviewwidget.cpp
#include <QPrintPreviewWidget>
#include <QPrinter>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QWidget>
#include <qprintpreviewwidget.h>
#include "libqprintpreviewwidget.h"
#include "libqprintpreviewwidget.hxx"

QPrintPreviewWidget* QPrintPreviewWidget_new(QWidget* parent) {
	 return new VirtualQPrintPreviewWidget(parent);
}

QPrintPreviewWidget* QPrintPreviewWidget_new2(QPrinter* printer) {
	 return new VirtualQPrintPreviewWidget(printer);
}

QPrintPreviewWidget* QPrintPreviewWidget_new3() {
	 return new VirtualQPrintPreviewWidget();
}

QPrintPreviewWidget* QPrintPreviewWidget_new4(QPrinter* printer, QWidget* parent) {
	 return new VirtualQPrintPreviewWidget(printer, parent);
}

QPrintPreviewWidget* QPrintPreviewWidget_new5(QPrinter* printer, QWidget* parent, int flags) {
	 return new VirtualQPrintPreviewWidget(printer, parent, static_cast<QFlags<Qt::WindowType>>(flags));
}

QPrintPreviewWidget* QPrintPreviewWidget_new6(QWidget* parent, int flags) {
	 return new VirtualQPrintPreviewWidget(parent, static_cast<QFlags<Qt::WindowType>>(flags));
}

libqt_string QPrintPreviewWidget_Tr(const char* s) {
	QString _ret = QPrintPreviewWidget::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

double QPrintPreviewWidget_ZoomFactor(const QPrintPreviewWidget* self) {
	return self->zoomFactor();
}

int QPrintPreviewWidget_Orientation(const QPrintPreviewWidget* self) {
	return self->orientation();
}

int QPrintPreviewWidget_ViewMode(const QPrintPreviewWidget* self) {
	return self->viewMode();
}

int QPrintPreviewWidget_ZoomMode(const QPrintPreviewWidget* self) {
	return self->zoomMode();
}

int QPrintPreviewWidget_CurrentPage(const QPrintPreviewWidget* self) {
	return self->currentPage();
}

int QPrintPreviewWidget_PageCount(const QPrintPreviewWidget* self) {
	return self->pageCount();
}

void QPrintPreviewWidget_SetVisible(QPrintPreviewWidget* self, bool visible) {
	self->setVisible(visible);
}

void QPrintPreviewWidget_Print(QPrintPreviewWidget* self) {
	self->print();
}

void QPrintPreviewWidget_ZoomIn(QPrintPreviewWidget* self) {
	self->zoomIn();
}

void QPrintPreviewWidget_ZoomOut(QPrintPreviewWidget* self) {
	self->zoomOut();
}

void QPrintPreviewWidget_SetZoomFactor(QPrintPreviewWidget* self, double zoomFactor) {
	self->setZoomFactor(zoomFactor);
}

void QPrintPreviewWidget_SetOrientation(QPrintPreviewWidget* self, int orientation) {
	self->setOrientation(static_cast<QPageLayout::Orientation>(orientation));
}

void QPrintPreviewWidget_SetViewMode(QPrintPreviewWidget* self, int viewMode) {
	self->setViewMode(static_cast<QPrintPreviewWidget::ViewMode>(viewMode));
}

void QPrintPreviewWidget_SetZoomMode(QPrintPreviewWidget* self, int zoomMode) {
	self->setZoomMode(static_cast<QPrintPreviewWidget::ZoomMode>(zoomMode));
}

void QPrintPreviewWidget_SetCurrentPage(QPrintPreviewWidget* self, int pageNumber) {
	self->setCurrentPage(pageNumber);
}

void QPrintPreviewWidget_FitToWidth(QPrintPreviewWidget* self) {
	self->fitToWidth();
}

void QPrintPreviewWidget_FitInView(QPrintPreviewWidget* self) {
	self->fitInView();
}

void QPrintPreviewWidget_SetLandscapeOrientation(QPrintPreviewWidget* self) {
	self->setLandscapeOrientation();
}

void QPrintPreviewWidget_SetPortraitOrientation(QPrintPreviewWidget* self) {
	self->setPortraitOrientation();
}

void QPrintPreviewWidget_SetSinglePageViewMode(QPrintPreviewWidget* self) {
	self->setSinglePageViewMode();
}

void QPrintPreviewWidget_SetFacingPagesViewMode(QPrintPreviewWidget* self) {
	self->setFacingPagesViewMode();
}

void QPrintPreviewWidget_SetAllPagesViewMode(QPrintPreviewWidget* self) {
	self->setAllPagesViewMode();
}

void QPrintPreviewWidget_UpdatePreview(QPrintPreviewWidget* self) {
	self->updatePreview();
}

void QPrintPreviewWidget_PaintRequested(QPrintPreviewWidget* self, QPrinter* printer) {
	self->paintRequested(printer);
}

void QPrintPreviewWidget_Connect_PaintRequested(QPrintPreviewWidget* self, intptr_t slot) {
    void (*slotFunc)(QPrintPreviewWidget*, QPrinter*) = reinterpret_cast<void (*)(QPrintPreviewWidget*, QPrinter*)>(slot);
    QPrintPreviewWidget::connect(self, &QPrintPreviewWidget::paintRequested, [self, slotFunc](QPrinter* printer) {
	slotFunc(self, printer);
    });
}

void QPrintPreviewWidget_PreviewChanged(QPrintPreviewWidget* self) {
	self->previewChanged();
}

void QPrintPreviewWidget_Connect_PreviewChanged(QPrintPreviewWidget* self, intptr_t slot) {
    void (*slotFunc)(QPrintPreviewWidget*) = reinterpret_cast<void (*)(QPrintPreviewWidget*)>(slot);
    QPrintPreviewWidget::connect(self, &QPrintPreviewWidget::previewChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

libqt_string QPrintPreviewWidget_Tr2(const char* s, const char* c) {
	QString _ret = QPrintPreviewWidget::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QPrintPreviewWidget_Tr3(const char* s, const char* c, int n) {
	QString _ret = QPrintPreviewWidget::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPrintPreviewWidget_ZoomIn1(QPrintPreviewWidget* self, double zoom) {
	self->zoomIn(zoom);
}

void QPrintPreviewWidget_ZoomOut1(QPrintPreviewWidget* self, double zoom) {
	self->zoomOut(zoom);
}

// Base class handler implementation
void QPrintPreviewWidget_QBaseSetVisible(QPrintPreviewWidget* self, bool visible) {
	auto* vqprintpreviewwidget = dynamic_cast<VirtualQPrintPreviewWidget*>(self);
	if (vqprintpreviewwidget && vqprintpreviewwidget->isVirtualQPrintPreviewWidget) {
vqprintpreviewwidget->setQPrintPreviewWidget_SetVisible_IsBase(true);
	vqprintpreviewwidget->setVisible(visible);
}
}

// Auxiliary method to allow providing re-implementation
void QPrintPreviewWidget_OnSetVisible(QPrintPreviewWidget* self, intptr_t slot) {
	auto* vqprintpreviewwidget = dynamic_cast<VirtualQPrintPreviewWidget*>(self);
	if (vqprintpreviewwidget && vqprintpreviewwidget->isVirtualQPrintPreviewWidget) {
vqprintpreviewwidget->setQPrintPreviewWidget_SetVisible_Callback(reinterpret_cast<VirtualQPrintPreviewWidget::QPrintPreviewWidget_SetVisible_Callback>(slot));
}
}

void QPrintPreviewWidget_Delete(QPrintPreviewWidget* self) {
    delete self;
}

